include/oneapi/tbb/collaborative_once_flag.h

```diff
@@ -0,0 +1,126 @@
+
+    ~run_once() {
+        // Caller is responsable for not pulling this out from under someone else.
+        // We could use a std::shared_mutex though...
+    }
+
+    [[nodiscard]] bool is_canceling() const { return my_context.is_group_execution_cancelled(); }
+
+    template <typename F>
+    task_group_status run_and_wait(F&& f) {
+        detail::d1::function_stack_task<F> t{ f, my_wait_context };
+        my_wait_context.reserve();
+        bool cancellation_status = false;
+        auto func = [&]() {
+            detail::try_call([&] { execute_and_wait(t, my_context, my_wait_context, my_context); }).on_completion([&] {
+                cancellation_status = my_context.is_group_execution_cancelled();
+                // Don't bother resetting the context. This isn't reusable.
+            });
+        };
+        my_arena.execute([&func]() { return tbb::this_task_arena::isolate(func); });
+        return cancellation_status ? canceled : complete;
+    }
+
+    /*! A call to wait() potentially participates in the work of run_and_wait().
+     * If run_and_wait() throws or is canceled, this returns canceled.
+     */
+    task_group_status wait() {
+        bool cancellation_status = false;
+        auto f = [&]() {
+            detail::try_call([&] {
+                detail::d1::wait(my_wait_context, my_context);
+            }).on_completion([&] {
+                cancellation_status = my_context.is_group_execution_cancelled();
+            });
+        };
+        my_arena.execute(f);
+        return cancellation_status ? canceled : complete;
+    }
+
+};
+
+} // namespace detail
+
+class canceled_exception : public std::exception {};
+
+// The concurrent_once_flag idea may be added to tbb at some point: oneapi-src/oneTBB#267
+class collaborative_once_flag : detail::d0::no_copy {
+    enum class state { uninitialized, in_progress, done };
+    std::atomic<state> my_state = state::uninitialized;
+    std::shared_ptr<detail::run_once> my_runner = nullptr;
+
+    template <typename Fn>
+    friend void collaborative_call_once(collaborative_once_flag& flag, Fn&& f);
+
+    template <typename Fn>
+    void do_collaborative_call_once(Fn&& f) {
+        while (my_state.load(std::memory_order_acquire) !=
+               state::done) { // Loop until someone else finishes or it's Done.
+            state expected = state::uninitialized;
+            if (my_state.compare_exchange_weak(expected, state::in_progress)) {
+                // We won the race to be the primary thread.
+                auto runner = std::make_shared<detail::run_once>();
+                std::atomic_store_explicit(&my_runner, runner, std::memory_order_release);
+                try {
+                    tbb::task_group_status wait_result = runner->run_and_wait(f);
+                    if (wait_result == tbb::task_group_status::canceled) {
+                        throw canceled_exception(); // Translate cancelation to an exception.
```

Comment:
Usually, exceptions are not recommended way for expected program flow while cancellation can be a part of normal program flow (e.g. looking for a element in container). I would propose to pass the context as an argument to `collaborative_call_once`.

Reply:
That seems reasonable! What would that look like in this case?

```
class LazyEvaluated {
    tbb::collaborative_once_flag;
    double cachedPropoerty;
public:
    double getProperty() {
        tbb::collaberative_call_once([]() {
             ... work ...
             cachedPropoerty = result;
        });
        return cachedProperty;
    }
};
```

`getProperty()` could conceivably could be called from different contexts at the same time? Would that be ok?

Reply:
`collaberative_call_once` is synchronization point that should provide an _acquire_ semantics. So, `cachedProperty` can be read concurrently without introducing a data race. In addition, the primary thread (that proceed the function until the end) should provide _release_ semantics.

Reply:
Hi Alexey, I was just wondering if the proper usage would be to pass the context into the getProperty() call, for it to be created on the stack in `getProperty()`, or if the LazyEvaluation class would have a context as a member?

Reply:
I think it should be similar to other parallel algorithms when the context is passed as the last optional argument. If the user needs something from the context, they can specify it as the last argument to `tbb::collaberative_call_once`. If not specified, a default context will be created on the stack. In addition, user can capture the explicit context into the lambda (e.g. for cancellation purposes).

Reply:
sounds good!

Reply:
@aleksei-fedotov in the case that a follow-on collaborator is called with a context and the first-to-arrive doesn't bail out prematurely, the behavior would be that the follow-on collaborator ignores its context, correct?

Reply:
> Usually, exceptions are not recommended way for expected program flow while cancellation can be a part of normal program flow (e.g. looking for a element in container). I would propose to pass the context as an argument to `collaborative_call_once`.

This is a good question. I've been bitten hard by cancelation causing me to think I've succeeded leading to UB (https://github.com/oneapi-src/oneTBB/issues/253). That said, the Committee has decided that `std::stop_token` shouldn't use exceptions for cancelation (early versions of `std::stop_token` had a `throw_if_canceled` member function). OTOH, `std::call_once` either succeeds or throws; there's no other way out. We could have it be
```
template <typename Fn>
[[nodiscard]] task_group_status collaborative_call_once(collaborative_once_flag& flag, Fn&& fn);
```
Then if client code wants it to throw on cancelation, we can easily wrap it. But if we do that, please please make it `[[nodiscard]]`!

Reply:
> @aleksei-fedotov in the case that a follow-on collaborator is called with a context and the first-to-arrive doesn't bail out prematurely, the behavior would be that the follow-on collaborator ignores its context, correct?

The context of first-to-arrive call is expected to be used for all collaborators. Overall, it seems strange if different contexts are used (or some threads do not specify the context). (P.S. I believe you wanted to notify @alexey-katranov).

As for exception vs cancellation, consider the following serial example:
```
try {
    bool zero_element_found = false;
    for (auto v : container) {
        if (!v.ptr) {
            throw bad_element;                         // exception on error
        }
        if (*v.ptr == 0) {
            zero_element_found = true;
            break;                                     // cancelation on normal program flow condition
        }
    }
    if (zero_element_found) { ... }
} catch ( ... ) {
    // report error
}
```
The main issue that neither exception nor cancelation will not work as expected in a threaded application because the both cancelation and exception are local to the executing thread. While the `break` condition is semi working because threads that met the condition will save some cycles, the exception condition can be dangerous because the exception will be thrown on some other thread that might miss sending a signal about the completion to the main thread.

oneTBB provides mechanisms to handle the both situations. The exception condition is handled automatically that any exception is captured and rethrown on the waiting thread, so it works like the exception occurs on the waiting thread. To improve the performance, the exception propagation mechanism triggers the cancelation mechanism that sets the flag into `tbb::task_group_context` that any not currently running task associated with this context should be skipped. However, we do not need to work with context (only to request the cancelation), i.e. the example can be transformed to parallel execution in the following way:
```
try {
    std::atomic<bool> zero_element_found = false;
    parallel_for (auto v : container) {
        if (!v.ptr) {
            throw bad_element;                         // exception on error
        }
        if (*v.ptr == 0) {
            zero_element_found = true;
            tbb::task::current_context()->cancel_group_execution();  // cancelation on normal program flow condition
         }
    }
    if (zero_element_found) { ... }
} catch ( ... ) {
    // report error
}
```
So, in some cases, you do not need to pass `task_group_context` explicitly to a functor or parallel algorithm.

The same idea can be applied to `collaborative_call_once` that any exception is processed like in a serial code and cancelation can be explicitly requested if necessary.

As `std::call_once` it has serial semantics and cancelation is not really required (user can use their own flags if necessary).